### PR TITLE
Use `cudf::detail::make_counting_transform_iterator`

### PR DESCRIPTION
**Pull request by codereport**
Came across some "raw" uses of `thrust::make_transform_iterator` + `thrust::make_counting_iterator` where now `cudf::detail::make_counting_transform_iterator` can be used (because of https://github.com/rapidsai/cudf/pull/6546).

### CODE REVIEW

**Comment by codereport** [comment]
rerun tests

**Review by devavret — APPROVED** [review]

File: cpp/include/cudf/detail/iterator.cuh [hunk 1]
```diff
@@ -43,6 +43,32 @@
...
+ * iter[1] == 1
+ * iter[2] == 4
+ * ...
+ * iter[n] == n * n
+ * @endcode
+ *
+ * @param start The starting value of the counting iterator
+ * @param f The unary function to apply to the counting iterator.
+ * This should be a host function and not a device function.
+ * @return auto A transform iterator that applies `f` to a counting iterator
```
Comment (davidwendt) [inline cpp/include/cudf/detail/iterator.cuh:64]:
```suggestion
 * @return A transform iterator that applies `f` to a counting iterator
```

File: cpp/include/cudf/detail/iterator.cuh [hunk 2]
```diff
@@ -43,6 +43,32 @@
...
+ * iter[0] == 0
+ * iter[1] == 1
+ * iter[2] == 4
+ * ...
+ * iter[n] == n * n
+ * @endcode
+ *
+ * @param start The starting value of the counting iterator
+ * @param f The unary function to apply to the counting iterator.
+ * This should be a host function and not a device function.
```
Comment (davidwendt) [inline cpp/include/cudf/detail/iterator.cuh:63]: Doesn't this depend on how the iterator is used? It certainly is being used with device functors in this PR.

**Review by davidwendt — CHANGES_REQUESTED** [review]

File: cpp/include/cudf/detail/iterator.cuh (hunk 2 above)
Comment (codereport) [reply cpp/include/cudf/detail/iterator.cuh:63]: That is a good point. @jrhemstad is there a reason it says this?

Comment (jrhemstad) [reply cpp/include/cudf/detail/iterator.cuh:63]: I think this was specific to when it was solely a test utility where the iterator was only ever invoked in host code.

File: cpp/include/cudf/detail/iterator.cuh [hunk 3]
```diff
@@ -254,8 +277,7 @@ auto inline make_scalar_iterator(scalar const& scalar_value)
 {
   CUDF_EXPECTS(data_type(type_to_id<Element>()) == scalar_value.type(), "the data type mismatch");
   CUDF_EXPECTS(scalar_value.is_valid(), "the scalar value must be valid");
-  return thrust::make_transform_iterator(thrust::make_constant_iterator<size_type>(0),
-                                         scalar_value_accessor<Element>{scalar_value});
+  return make_counting_transform_iterator(0, scalar_value_accessor<Element>{scalar_value});
```
Comment (harrism) [inline cpp/include/cudf/detail/iterator.cuh:280]: I don't like replacing transform(constant) with transform(counting). These are not the same thing when you get to the assembly code. I don't see a need for this change, or a benefit from it.

**Review by davidwendt — APPROVED** [review]

File: cpp/include/cudf/detail/iterator.cuh [hunk 4]
```diff
@@ -316,34 +338,7 @@ auto inline make_pair_iterator(scalar const& scalar_value)
 {
   CUDF_EXPECTS(type_id_matches_device_storage_type<Element>(scalar_value.type().id()),
                "the data type mismatch");
-  return thrust::make_transform_iterator(thrust::make_constant_iterator<size_type>(0),
```
Comment (harrism) [inline cpp/include/cudf/detail/iterator.cuh:319]: Another case like above. Suggest leaving this alone.

**Review by harrism — CHANGES_REQUESTED** [review]
Great, except the cases where you replace transform(constant) with transform(counting).

File: cpp/include/cudf/detail/iterator.cuh (hunk 3 above)
Comment (codereport) [reply cpp/include/cudf/detail/iterator.cuh:280]: Oh yea, this is a total mistake. Great catch! I am surprised CI didn't break.

**Review by harrism — APPROVED** [review]

**Comment by harrism** [comment]
@gpucibot merge

**Pull request merged**